pkg/services/ngalert/api/api_prometheus_test.go

```diff
@@ -414,18 +418,59 @@ func TestRouteGetRuleStatuses(t *testing.T) {
 }
 `, folder.Title), string(r.Body()))
 	})
+
+	t.Run("when fine-grained access is enabled", func(t *testing.T) {
+		t.Run("should return only rules the user can query all data sources", func(t *testing.T) {
+			ruleStore := store.NewFakeRuleStore(t)
+			fakeAIM := NewFakeAlertInstanceManager(t)
+
+			rules := ngmodels.GenerateAlertRules(rand.Intn(4)+2, ngmodels.AlertRuleGen(withOrgID(orgID)))
+			ruleStore.PutRule(context.Background(), rules...)
+			ruleStore.PutRule(context.Background(), ngmodels.GenerateAlertRules(rand.Intn(4)+2, ngmodels.AlertRuleGen(withOrgID(orgID)))...)
```

Comment:
Why randomize the number of rules in the test?

Generally I prefer to reduce randomness in tests wherever possible (aside from fuzz tests). If one of the random results actually causes the test to fail, it won't be easy to reproduce/fix locally.

Reply:
In this particular case, a random number of alert rules is not really necessary as the test does not check them. I just copy-pasted lines and followed the general idea of writing tests that I've applied so far.

The idea behind randomizing input data is to emphasize that some specific aspects of input data do not matter for that particular test as well as spotlight those that do matter by specifying them with constants or limiting the range of random values that is valid for the specific test. 

When a test tests a method that validates a specific field of a structure that has many fields besides the one under the test, it is generally a good idea to have other fields to be populated with random data rather than some constant or (worse) default. Randomizing input data creates a useful noise and makes the method under test to be more reliable as well as tests be more fragile. The fragility of the test is actually a good feature - if the method is changed and starts using the input data in some different way (new fields, or a limited number of elements in the input array, etc), a good test will likely to get broken or at least become flacky. 

> If one of the random results actually causes the test to fail, it won't be easy to reproduce/fix locally.

This means that either the test is incorrectly written or the method under test has been changed. 

> it won't be easy to reproduce/fix locally.

This is a good point. However, I do not think that should stop us from using randomization. To be able to reproduce the randomized input data we should log the randomizer seed. That will make tests reproducible.